AlphaESS Integration

CharlesGillanders opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
--

## Proposed change
<!--
  Adds integration to AlphaESS energy storage solutions
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20868

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

CharlesGillanders committed:
First commit - AlphaESS integration

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    def __init__(self, hass: HomeAssistant, client: alphaess.alphaess) -> None:
+        """Initialize."""
+        self.api = client
+        self.platforms = ["sensor"]
```
Why is this required? (and why not reuse the constant from `__init__`, or better move it to const.py`)

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        self.platforms = ["sensor"]
+        self.update_method = self._async_update_data
+
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
```
I think that best practice is to call `super().__init__` as the first thing, but someone with more experience should verify this.

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        try:
+            _LOGGER.info("Trying to query AlphaESS api data")
+            return await self.api.getdata()
+        except Exception as exception:
```
Do you have specific exceptions in your client class? Or does it always throw a generic exception, where it can recover from if you try again?

CharlesGillanders replied:
The client class is not yet very sophisticated.  I know it needs more work to get to a state where the home assistant integration can move up the quality scale.

iMicknl replied:
Those broad exceptions aren't allowed most of the time, so I would advise to see if you can catch errors like TimeOuts or wrong credentials asap.

CharlesGillanders replied:
Taken me too long to come back to this but I have updated the client class to identify connection errors and authentication problems separately.

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)
```

iMicknl commented on homeassistant/components/alphaess/strings.json:
```diff
@@ -0,0 +1,20 @@
+{
+  "config": {
+    "flow_title": "Alpha ESS: {username}",
```
I don't think this extra username hint is required. I would just remove the flow title to keep it simple.

iMicknl commented on homeassistant/components/alphaess/strings.json:
```diff
@@ -0,0 +1,20 @@
+    },
+    "step": {
+        "user": {
+            "title": "Set up Alpha ESS",
```
Remove the title, just keep it the default.

iMicknl commented on homeassistant/components/alphaess/strings.json:
```diff
@@ -0,0 +1,20 @@
+    "step": {
+        "user": {
+            "title": "Set up Alpha ESS",
+            "description": "Please provide your credentials for logging into the Alpha ESS app/website",
```
```suggestion
            "description": "Please provide your credentials for logging into the Alpha ESS app/website.",
```

iMicknl commented on homeassistant/components/alphaess/manifest.json:
```diff
@@ -0,0 +1,17 @@
+  "requirements": [
+    "alphaess==0.0.2"
+  ],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
```suggestion
```

Remove unused keys

CharlesGillanders replied:
Thanks - I've removed those unused keys.

iMicknl commented on homeassistant/components/alphaess/const.py:
```diff
@@ -0,0 +1,10 @@
+
+
+NAME = "Alpha ESS"
+VERSION = "0.0.2"
```
```suggestion
```
Why do you need this? It seems not used.

iMicknl commented on homeassistant/components/alphaess/const.py:
```diff
@@ -0,0 +1,10 @@
+
+NAME = "Alpha ESS"
+VERSION = "0.0.2"
+ISSUE_URL = "https://github.com/CharlesGillanders/homeassistant-alphaESS/issues"
```
```suggestion
```
Remove

iMicknl commented on homeassistant/components/alphaess/const.py:
```diff
@@ -0,0 +1,10 @@
+DOMAIN = "alphaess"
+
+
+NAME = "Alpha ESS"
```
Why do you need this? (seems unused)
```suggestion
```

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        hass.data.setdefault(DOMAIN, {})
+
+    client = alphaess.alphaess()
+    client.username = entry.data.get("username")
```
```suggestion
    client.username = entry.data[CONF_USERNAME]
```

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+
+    client = alphaess.alphaess()
+    client.username = entry.data.get("username")
+    client.password = entry.data.get("password")
```
```suggestion
    client.password = entry.data[CONF_PASSWORD]
```

iMicknl reviewed (changes requested):
Thanks! I had a quick look already and had some remarks. Didn't check config flow and sensor yet, but will have a look after you addressed some of the feedback.

CharlesGillanders commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+# from homeassistant.helpers.aiohttp_client import async_get_clientsession
```
Thanks for spotting that incorrect commented code.  I'll remove it.

CharlesGillanders commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+
+# TODO List the platforms that you want to support.
+# For your initial PR, limit it to 1 platform.
+PLATFORMS = ["sensor"]
```
Thanks for the tip, I've replaced the string with this enum.

CharlesGillanders commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Alpha ESS from a config entry."""
+
+    if hass.data.get(DOMAIN) is None:
+        hass.data.setdefault(DOMAIN, {})
```
Thanks - I've removed that.

CharlesGillanders commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    coordinator = AlphaESSDataUpdateCoordinator(hass, client=client)
+    await coordinator.async_config_entry_first_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
```
Again - thank you.  I've removed the unnecessary code.

CharlesGillanders commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    return True
+
+
+class AlphaESSDataUpdateCoordinator(DataUpdateCoordinator):
```
I've made that change - thanks for the suggestion.

CharlesGillanders committed:
Updates requested for PR 62801

iMicknl commented on tests/components/alphaess/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+"""Test the AlphaEss config flow."""
```
Since you only have tests for the config flow, you need to exclude the others in .codecoveragerc

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,127 @@
+            if self._serial == serial:
+                model = invertor["minv"]
+
+        self._attr_device_info = {
+            ATTR_IDENTIFIERS: {(DOMAIN, serial)},
+            ATTR_NAME: f"Alpha ESS Energy Statistics : {serial}",
+            ATTR_MANUFACTURER: "AlphaESS",
+            ATTR_MODEL: model,
+            ATTR_ENTRY_TYPE: DeviceEntryType.SERVICE,
+        }
```
```suggestion
        self._attr_device_info = DeviceInfo(
            entry_type=DeviceEntryType.SERVICE,
            identifiers={(DOMAIN, serial)},
            manufacturer="AlphaESS",
            model=model,
            name= f"Alpha ESS Energy Statistics : {serial}",
        )
```

Use the typed device info from `from homeassistant.helpers.entity import DeviceInfo`

CharlesGillanders replied:
Thanks for the advice!  Much appreciated.

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        serial = invertor["sys_sn"]
+        async_add_entities(
+            [
+                AlphaESSSensor(coordinator, entry, serial, "Solar Production"),
```
Did you look at creating `SensorEntityDescription`? This will help you define the entities in a nice way, where you don't have to add a lot of if statements in your actual sensor implementation.

See tuya/overkiz/shelly platforms for an example.

CharlesGillanders replied:
I've just made the update to use SensorEntityDescription.  Thanks for the examples.

iMicknl commented on homeassistant/components/alphaess/const.py:
```diff
@@ -0,0 +1,5 @@
+
+DOMAIN = "alphaess"
+
+ATTR_ENTRY_TYPE = "entry_type"
```
Remove. (see DeviceInfo comment later on)

iMicknl commented on tests/components/alphaess/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+    assert result["errors"] is None
+
+    with patch(
+        "homeassistant.components.alphaess.config_flow.PlaceholderHub.authenticate",
```
Where is this PlaceholderHub class? It seems that your tests are not written yet, just the default tests.

Please make sure to add Config Flow coverage, since this is required for core.

CharlesGillanders replied:
I have written the tests for config flow coverage - I think this is complete now.

iMicknl commented on homeassistant/components/alphaess/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+        try:
+            _LOGGER.info("Trying to query AlphaESS api data")
+            return await self.api.getdata()
+        except Exception as exception:
```
Remove broad except, DataUpdateCoordinator already handles those broad excepts.

CharlesGillanders replied:
Fixed thank you - I know separately catch client response errors and connection errors and leave all else to the DataUpdateCoordinator

iMicknl commented on homeassistant/components/alphaess/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+            raise UpdateFailed() from exception
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Shouldn't this be part of `__init__`?

CharlesGillanders replied:
Yes it should be, my mistake.

iMicknl commented on homeassistant/components/alphaess/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+from .const import DOMAIN
+
+_LOGGER: logging.Logger = logging.getLogger(__package__)
+SCAN_INTERVAL = timedelta(minutes=5)
+PLATFORMS = [Platform.SENSOR]
```
Move those parts to const or `__init__`.

CharlesGillanders replied:
I have moved them to const

iMicknl commented on homeassistant/components/alphaess/config_flow.py:
```diff
@@ -0,0 +1,75 @@
+
+        try:
+            await validate_input(self.hass, user_input)
+        except CannotConnect:
```
You are catching this, but never throwing this?

CharlesGillanders replied:
Thank you - removed the unused exception and replaced with exceptions that I do use from the client class.

iMicknl reviewed (changes requested):
Thanks for addressing the changes. I have made some follow up remarks. Unfortunately I cannot kick off the CI/CD, but I think that pytest and mypy will fail as well.

CharlesGillanders committed:
Use updated alphaess pypy library version 0.0.3 (supports more detailed exceptions)

Addressed pull comments
https://github.com/home-assistant/core/pull/62801#discussion_r776386569
https://github.com/home-assistant/core/pull/62801#discussion_r776398169
https://github.com/home-assistant/core/pull/62801#discussion_r776398806
https://github.com/home-assistant/core/pull/62801#discussion_r776399796

CharlesGillanders committed:
Updated for https://github.com/home-assistant/core/pull/62801#discussion_r776397769

CharlesGillanders committed:
added unit tests for config_flow

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+# class AlphaESSInverter():
+#     """Class for Inverter data."""
+
+#     def __init__(self, SolarProduction,SolarToBattery,SolarToGrid,SolarToLoad,TotalLoad,GridToLoad,GridToBattery,StateOfCharge,Charge,Discharge):
+#         self.SolarProduction = SolarProduction
+#         self.SolartToBattery = SolarToBattery
+#         self.SolarToGrid = SolarToGrid
+#         self.SolarToLoad = SolarToLoad
+#         self.TotalLoad = TotalLoad
+#         self.GridToLoad = GridToLoad
+#         self.GridToBattery = GridToBattery
+#         self.StateOfCharge = StateOfCharge
+#         self.Charge = Charge
+#         self.Discharge = Discharge
```
```suggestion
```

CharlesGillanders replied:
Fixed.

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    return unload_ok
+
+
+async def async_update_options(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Update options."""
+    await hass.config_entries.async_reload(entry.entry_id)
```
```suggestion
```

You are not using the OptionsFlow in your ConfigFlow? So this is not required.

iMicknl commented on homeassistant/components/alphaess/__init__.py:
```diff
@@ -0,0 +1,59 @@
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    entry.async_on_unload(entry.add_update_listener(async_update_options))
+
```
```suggestion
```

iMicknl commented on homeassistant/components/alphaess/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    client = alphaess.alphaess()
+
+    try:
+        await client.authenticate(data["username"], data["password"])
```
```suggestion
        await client.authenticate(data[CONF_USERNAME], data[CONF_PASSWORD])
```
Please use the constants everywhere, there are some more ocassions.

iMicknl commented on homeassistant/components/alphaess/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        if e.status == 401:
+            raise InvalidAuth
+        else:
+            raise UnknownError
```
```suggestion
            raise UnknownError from e
```

It would be good to understand what the error is. You can also just `raise e` to reraise the original exception.

CharlesGillanders replied:
Thank you - that is a better result.

iMicknl commented on homeassistant/components/alphaess/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    except aiohttp.client_exceptions.ClientConnectorError:
+        raise CannotConnect
+
+    else:
+        return {"AlphaESS": data["username"]}
```
```suggestion
    return {"AlphaESS": data["username"]}
```

iMicknl commented on homeassistant/components/alphaess/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+            errors["base"] = "invalid_auth"
+        except UnknownError:
+            errors["base"] = "unknown"
+        else:
+            return self.async_create_entry(
+                title=user_input["username"], data=user_input
+            )
```
Remove the else. Why do you use an `else` after a try except?

iMicknl commented on homeassistant/components/alphaess/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
```
Instead of this duplicate code, you could also wrap the other parts in

```suggestion
        if user_input:
```

iMicknl commented on homeassistant/components/alphaess/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+    async def _async_update_data(self):
+        """Update data via library."""
+        try:
+            _LOGGER.info("Trying to query AlphaESS api data")
```
```suggestion
```
Why do you need this info statement? Doesn't add anything imo.

iMicknl commented on homeassistant/components/alphaess/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+            for invertor in jsondata:
+                index = int(datetime.date.today().strftime("%d")) - 1
+                inverterdata: dict[str, any] = {}
+                inverterdata.update({"Model": invertor["minv"]})
+                inverterdata.update(
+                    {"Solar Production": invertor["statistics"]["EpvT"]}
+                )
+                inverterdata.update(
+                    {"Solar to Battery": invertor["statistics"]["Epvcharge"]}
+                )
+                inverterdata.update({"Solar to Grid": invertor["statistics"]["Eout"]})
+                inverterdata.update(
+                    {"Solar to Load": invertor["statistics"]["Epv2load"]}
+                )
+                inverterdata.update({"Total Load": invertor["statistics"]["EHomeLoad"]})
+                inverterdata.update(
+                    {"Grid to Load": invertor["statistics"]["EGrid2Load"]}
+                )
+                inverterdata.update(
+                    {"Grid to Battery": invertor["statistics"]["EGridCharge"]}
+                )
+                inverterdata.update({"State of Charge": invertor["statistics"]["Soc"]})
+                inverterdata.update(
+                    {"Charge": invertor["system_statistics"]["ECharge"][index]}
+                )
+                inverterdata.update(
+                    {"Discharge": invertor["system_statistics"]["EDischarge"][index]}
+                )
```
Why do you hardcode the mapping here to `inverterdata`? Can you not use the key (like `EDischarge`) as key in your entity description? This will remove a lot of (unnecessary) overhead.

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    for invertor in coordinator.data:
+        serial = invertor
```
Why is this required?

CharlesGillanders replied:
This is the same situation as I note in https://github.com/home-assistant/core/pull/62801#discussion_r814142234  basically AlphaESS supports someone owning multiple inverters accessed under a single login to their API/website.  I want to make sure that I query all of the inverters where there are multiples.  Is there a better more home assistant best practice way of doing this?

iMicknl replied:
In this case I meant why do you need to assign invertor to serial directly. Just say `for serial in coordinator.data` otherwise.

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        """Return the name of the sensor."""
+        return f"{self._serial} - {self._name}"
+
+    @property
+    def native_unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        key_supported_states = {
+            description.key: description for description in SENSOR_DESCRIPTIONS
+        }
+        return key_supported_states[self._name].native_unit_of_measurement
```
```suggestion
```
This is not how it works, you can just remove this and it will be set via the entity description.

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        for description in key_supported_states:
+            entities.append(
+                AlphaESSSensor(
+                    coordinator, entry, serial, key_supported_states[description].name
```
```suggestion
                    coordinator, entry, serial, description
```

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._config = config
+        self._name = name
```
Don't pass the name, but set the description. Have a look at tuya e.g. for more info.

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,160 @@
+    @property
+    def unique_id(self):
+        """Return a unique ID to use for this entity."""
+        return f"{self._config.entry_id}_{self._serial} - {self._name}"
```
I don't think you should use spaces here. Is serial not unique enough? (maybe together with the key)

CharlesGillanders replied:
I could use some advice here - AlphaESS supports the idea of an individual owning multiple inverters and solar energy setups.  The serial number is in reference to a specific inverter and then each inverter has multiple sensors which can be queried for various statistics - Solar2Battery, Solar2Load, BatteryState etcetera.  If I just include the serial then how do I differentiate between the various individual sensors.  If I don't include the serial then how do I differentiate between the same sensor for different inverters?  Happy to take best practice for home assistant and try to implement it.

iMicknl replied:
If serial is unique per physical device, I would indeed just use `{self.serial}_{self.name}`. 

You can use serial as unique id for a device (not entity), to link all sensors together. Where all entities can use the combination of serial and name.

iMicknl commented on homeassistant/components/alphaess/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        self._serial = serial
+        self._coordinator = coordinator
+
+        for invertor in coordinator.data:
+            serial = invertor
+            if self._serial == serial:
+                self._attr_device_info = DeviceInfo(
+                    entry_type=DeviceEntryType.SERVICE,
+                    identifiers={(DOMAIN, serial)},
+                    manufacturer="AlphaESS",
+                    model=coordinator.data[invertor]["Model"],
+                    name=f"Alpha ESS Energy Statistics : {serial}",
+                )
```
Why do you need this for every sensor?

iMicknl reviewed (changes requested):
Some more comments to get you started. Thanks @CharlesGillanders
